### PR TITLE
GC OS interface refactoring

### PR DESCRIPTION
**Pull request by janvorli**
This change replaces all calls of OS specific functions in the GC by a call to
a platform agnostic interface. Critical sections were abstracted too.
The logging file access was changed to use CRT functions instead of Windows specific APIs.
A "size" member was added to the card_table_info so that we can pass the right
size to the VirtualRelease method when destroying the card table.
I have also fixed a bug in the gc_heap::make_card_table error path where when VirtualCommit
failed, it called VirtualRelease with size that was not the reserved size, but
the committed size.

### CODE REVIEW

**Comment by janvorli** [comment]
Please note that github doesn't show diff for the file with the largest amount of changes, the gc.cpp.
@jkotas Can you take a look please?

**Comment by janvorli** [comment]
cc: @smosier

File: src/vm/gcenv.cpp [hunk 1]
```diff
@@ -694,3 +694,345 @@ Thread * GCToEEInterface::GetThreadList(Thread * pThread)
...
+void GCToOSInterface::Shutdown()
+{
+}
+
+// Create a new lock
+// Return:
+//   An opaque handle representing the lock
+void* GCToOSInterface::CreateLock()
+{
+    CRITICAL_SECTION* cs = new (nothrow) CRITICAL_SECTION;
```
Comment (jkotas) [inline src/vm/gcenv.cpp]: null check

**Comment by janvorli** [comment]
@dotnet-bot test OSX x64 Debug

**Comment by janvorli** [comment]
@jkotas - I have pushed a new commit:
- All interlocked operations moved to Interlocked class as static methods
- Removed unused function prototypes
- Shuffled stuff in the root CMakeLists.txt to enable building the GC sample using the
  settings inherited from the root CMakeLists.txt and to clean up some things that have
  rotted over time, like the FEATURE_xxx macros not being in one alphabetically ordered
  block
- Fixed the VOLATILE_MEMORY_BARRIER macro in the gcenv.base.h
- Replaced uint32_t thread id by EEThreadId
- Removed thread handles storage (g_gc_thread) from the GC. The thread handle is closed right after the thread is launched. That allowed me to get rid of the GCThreadHandle
- Renamed the methods of the EEThreadId to be easier to understand
- Moved the gcenv.windows.cpp and gcenv.unix.cpp to the sample folder

**Pull request merged**